Error in requestLoop() java.lang.NullPointerException

Sorry, this is going to sound really vague but I get the following stacktrace pretty frequently on a server when (I assume) it accepts `PUT` requests.

I don't have much to go on here so thought I'd post an issue for a few pointers to help me debug. The server is pretty simple, started up like below and runs just a few `HttpService` endpoints.

```
  private def build(): Task[org.http4s.server.Server] = BlazeBuilder
    .withServiceExecutor(DefaultExecutorService)
    .bindHttp(port, "0.0.0.0")
    .mountService(services(), "/")
    .start
```

The `PUT` sends a simple JSON body. Not sure what else I can tell you...

The stack trace is below

```
[blaze-nio-fixed-selector-pool-2] ERROR org.http4s.blaze.pipeline.Stage - error in requestLoop()
java.lang.NullPointerException
	at org.http4s.internal.parboiled2.ParserInput$StringBasedParserInput.length(ParserInput.scala:97)
	at org.http4s.internal.parboiled2.Parser.__advance(Parser.scala:228)
	at org.http4s.internal.parboiled2.Parser.runRule$1(Parser.scala:140)
	at org.http4s.internal.parboiled2.Parser.phase0_initialRun$1(Parser.scala:150)
	at org.http4s.internal.parboiled2.Parser.__run(Parser.scala:203)
	at org.http4s.parser.Rfc2616BasicRules$.token(Rfc2616BasicRules.scala:84)
	at org.http4s.Method$.$anonfun$fromString$1(Method.scala:47)
	at scala.collection.MapLike.getOrElse(MapLike.scala:128)
	at scala.collection.MapLike.getOrElse$(MapLike.scala:126)
	at scala.collection.concurrent.TrieMap.getOrElse(TrieMap.scala:631)
	at org.http4s.Method$.fromString(Method.scala:47)
	at org.http4s.server.blaze.Http1ServerParser.collectMessage(Http1ServerParser.scala:48)
	at org.http4s.server.blaze.Http1ServerStage.runRequest(Http1ServerStage.scala:111)
	at org.http4s.server.blaze.Http1ServerStage.reqLoopCallback(Http1ServerStage.scala:97)
	at org.http4s.server.blaze.Http1ServerStage.$anonfun$requestLoop$1(Http1ServerStage.scala:74)
	at org.http4s.server.blaze.Http1ServerStage.$anonfun$requestLoop$1$adapted(Http1ServerStage.scala:74)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at org.http4s.blaze.util.Execution$$anon$1.execute(Executor.scala:27)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)
	at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1(Promise.scala:284)
	at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1$adapted(Promise.scala:284)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
	at org.http4s.blaze.channel.nio1.NIO1HeadStage.readReady(NIO1HeadStage.scala:61)
	at org.http4s.blaze.channel.nio1.SelectorLoop.run(SelectorLoop.scala:134)
```

I switched to Jetty to see if it made a difference. I can't tell if this is the same underlying problem or not, but I now get this occasional error. Can you advice if this is the same issue? Any thoughts? thx

```
Tue 09-Jan-201821:19:59.365[temperature-http-server-1]ERRORError processing request after response was committed
org.eclipse.jetty.io.EofException: null
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:292) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:429) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:322) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:372) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.server.HttpConnection$SendCallback.process(HttpConnection.java:794) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.server.HttpConnection.send(HttpConnection.java:528) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:783) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:834) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:234) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.server.HttpOutput$AsyncWrite.process(HttpOutput.java:1205) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) ~[temperature-machine-2.1.jar:2.1]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:465) ~[temperature-machine-2.1.jar:2.1]
	at java.io.OutputStream.write(OutputStream.java:75) ~[?:1.8.0_65]
	at org.http4s.servlet.NonBlockingServletIo.$anonfun$initWriter$3(ServletIo.scala:187) ~[temperature-machine-2.1.jar:2.1]
	at org.http4s.servlet.NonBlockingServletIo.$anonfun$initWriter$3$adapted(ServletIo.scala:183) ~[temperature-machine-2.1.jar:2.1]
	at org.http4s.servlet.NonBlockingServletIo.$anonfun$initWriter$10(ServletIo.scala:243) ~[temperature-machine-2.1.jar:2.1]
	at org.http4s.servlet.NonBlockingServletIo.$anonfun$initWriter$10$adapted(ServletIo.scala:243) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Task.$anonfun$map$3(Task.scala:42) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Task$.Try(Task.scala:457) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Task.$anonfun$map$2(Task.scala:42) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.$bslash$div.flatMap(Either.scala:135) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Task.$anonfun$map$1(Task.scala:42) ~[temperature-machine-2.1.jar:2.1]
	at scala.Function1.$anonfun$andThen$1(Function1.scala:52) ~[temperature-machine-2.1.jar:2.1]
	at scala.Function1.$anonfun$andThen$1(Function1.scala:52) ~[temperature-machine-2.1.jar:2.1]
	at scala.Function1.$anonfun$andThen$1(Function1.scala:52) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future.$anonfun$unsafePerformListen$2(Future.scala:79) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Trampoline$.$anonfun$delay$1(Free.scala:356) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free$.$anonfun$suspend$2(Free.scala:27) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.$anonfun$resume$2(Free.scala:119) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.std.FunctionInstances$$anon$1.$anonfun$map$1(Function.scala:77) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.$anonfun$run$1(Free.scala:271) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.go2$1(Free.scala:162) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.go(Free.scala:165) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.run(Free.scala:271) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$.$anonfun$async$2(Future.scala:428) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$.$anonfun$async$2$adapted(Future.scala:428) ~[temperature-machine-2.1.jar:2.1]
	at org.http4s.servlet.NonBlockingServletIo.$anonfun$initWriter$9(ServletIo.scala:237) ~[temperature-machine-2.1.jar:2.1]
	at org.http4s.servlet.NonBlockingServletIo.$anonfun$initWriter$9$adapted(ServletIo.scala:232) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$.$anonfun$async$1(Future.scala:428) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$.$anonfun$async$1$adapted(Future.scala:428) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future.unsafePerformListen(Future.scala:79) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future.$anonfun$unsafePerformListen$3(Future.scala:79) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future.$anonfun$unsafePerformListen$3$adapted(Future.scala:79) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.$anonfun$map$1(Free.scala:99) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.resume(Free.scala:117) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.go2$1(Free.scala:161) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.go(Free.scala:165) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.run(Free.scala:271) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$$anon$3.call(Future.scala:432) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$$anon$3.call(Future.scala:432) ~[temperature-machine-2.1.jar:2.1]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_65]
	at org.http4s.util.TrampolineExecutionContext$.execute(TrampolineExecutionContext.scala:27) ~[temperature-machine-2.1.jar:2.1]
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134) ~[?:1.8.0_65]
	at scalaz.concurrent.Future$.$anonfun$apply$1(Future.scala:432) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$.$anonfun$apply$1$adapted(Future.scala:431) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future.unsafePerformListen(Future.scala:79) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future.$anonfun$unsafePerformListen$3(Future.scala:79) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future.$anonfun$unsafePerformListen$3$adapted(Future.scala:79) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.$anonfun$map$1(Free.scala:99) ~[temperature-machine-2.1.jar:2.1]
	at scalaz.Free.resume(Free.scala:117) [temperature-machine-2.1.jar:2.1]
	at scalaz.Free.go2$1(Free.scala:161) [temperature-machine-2.1.jar:2.1]
	at scalaz.Free.go(Free.scala:165) [temperature-machine-2.1.jar:2.1]
	at scalaz.Free.run(Free.scala:271) [temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$$anon$3.call(Future.scala:432) [temperature-machine-2.1.jar:2.1]
	at scalaz.concurrent.Future$$anon$3.call(Future.scala:432) [temperature-machine-2.1.jar:2.1]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_65]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_65]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_65]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_65]
Caused by: java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.writev0(Native Method) ~[?:1.8.0_65]
	at sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:51) ~[?:1.8.0_65]
	at sun.nio.ch.IOUtil.write(IOUtil.java:148) ~[?:1.8.0_65]
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:504) ~[?:1.8.0_65]
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:272) ~[temperature-machine-2.1.jar:2.1]
	... 71 more
```

--- a/blaze-server/src/main/scala/org/http4s/server/blaze/Http1ServerStage.scala
+++ b/blaze-server/src/main/scala/org/http4s/server/blaze/Http1ServerStage.scala
@@ -9,6 +9,7 @@ import java.nio.ByteBuffer
 import org.http4s.blaze.http.http_parser.BaseExceptions.{BadRequest, ParserException}
 import org.http4s.blaze.pipeline.Command.EOF
 import org.http4s.blaze.pipeline.{TailStage, Command => Cmd}
+import org.http4s.blaze.util.BufferTools
 import org.http4s.blaze.util.BufferTools.emptyBuffer
 import org.http4s.blaze.util.Execution._
 import org.http4s.blazecore.Http1Stage
@@ -59,16 +60,24 @@ private[blaze] class Http1ServerStage[F[_]](
 
   // micro-optimization: unwrap the service and call its .run directly
   private[this] val serviceFn = service.run
+
+  // both `parser` and `isClosed` are protected by synchronization on `parser`
   private[this] val parser = new Http1ServerParser[F](logger, maxRequestLineLen, maxHeadersLen)
+  private[this] var isClosed = false
 
   val name = "Http4sServerStage"
 
   logger.trace(s"Http4sStage starting up")
 
   final override protected def doParseContent(buffer: ByteBuffer): Option[ByteBuffer] =
-    parser.doParseContent(buffer)
+    parser.synchronized {
+      parser.doParseContent(buffer)
+    }
 
-  final override protected def contentComplete(): Boolean = parser.contentComplete()
+  final override protected def contentComplete(): Boolean =
+    parser.synchronized {
+      parser.contentComplete()
+    }
 
   // Will act as our loop
   override def stageStartup(): Unit = {
@@ -76,45 +85,53 @@ private[blaze] class Http1ServerStage[F[_]](
     requestLoop()
   }
 
-  private def requestLoop(): Unit = channelRead().onComplete(reqLoopCallback)(trampoline)
-
-  private def reqLoopCallback(buff: Try[ByteBuffer]): Unit = buff match {
-    case Success(buff) =>
-      logger.trace {
-        buff.mark()
-        val sb = new StringBuilder
-        while (buff.hasRemaining) sb.append(buff.get().toChar)
-
-        buff.reset()
-        s"Received request\n${sb.result}"
-      }
-
-      try {
-        if (!parser.requestLineComplete() && !parser.doParseRequestLine(buff)) {
-          requestLoop()
-          return
-        }
-        if (!parser.headersComplete() && !parser.doParseHeaders(buff)) {
-          requestLoop()
-          return
-        }
-        // we have enough to start the request
-        runRequest(buff)
-      } catch {
-        case t: BadRequest =>
-          badMessage("Error parsing status or headers in requestLoop()", t, Request[F]())
-        case t: Throwable =>
-          internalServerError(
-            "error in requestLoop()",
-            t,
-            Request[F](),
-            () => Future.successful(emptyBuffer))
-      }
-
+  private val handleReqRead: Try[ByteBuffer] => Unit = {
+    case Success(buff) => reqLoopCallback(buff)
     case Failure(Cmd.EOF) => stageShutdown()
     case Failure(t) => fatalError(t, "Error in requestLoop()")
   }
 
+  private def requestLoop(): Unit = channelRead().onComplete(handleReqRead)(trampoline)
+
+  private def reqLoopCallback(buff: ByteBuffer): Unit = {
+    logRequest(buff)
+    parser.synchronized {
+      if (!isClosed) {
+        try {
+          if (!parser.requestLineComplete() && !parser.doParseRequestLine(buff)) {
+            requestLoop()
+          } else if (!parser.headersComplete() && !parser.doParseHeaders(buff)) {
+            requestLoop()
+          } else {
+            // we have enough to start the request
+            runRequest(buff)
+          }
+        } catch {
+          case t: BadRequest =>
+            badMessage("Error parsing status or headers in requestLoop()", t, Request[F]())
+          case t: Throwable =>
+            internalServerError(
+              "error in requestLoop()",
+              t,
+              Request[F](),
+              () => Future.successful(emptyBuffer))
+        }
+      } else {
+        logger.debug(s"Parser closed. Buffer size: ${buff.remaining}")
+      }
+    }
+  }
+
+  private def logRequest(buffer: ByteBuffer): Unit =
+    logger.trace {
+      val msg = BufferTools
+        .bufferToString(buffer.duplicate())
+        .replace("\r", "\\r")
+        .replace("\n", "\\n\n")
+      s"Received Request:\n$msg"
+    }
+
+  // Only called while holding the monitor of `parser`
   private def runRequest(buffer: ByteBuffer): Unit = {
     val (body, cleanup) = collectBodyFromParser(
       buffer,
@@ -211,12 +228,12 @@ private[blaze] class Http1ServerStage[F[_]](
             bodyCleanup().onComplete {
               case s @ Success(_) => // Serve another request
                 parser.reset()
-                reqLoopCallback(s)
+                handleReqRead(s)
 
               case Failure(EOF) => closeConnection()
 
               case Failure(t) => fatalError(t, "Failure in body cleanup")
-            }(directec)
+            }(trampoline)
           }
 
       case Left(EOF) =>
@@ -236,7 +253,10 @@ private[blaze] class Http1ServerStage[F[_]](
 
   override protected def stageShutdown(): Unit = {
     logger.debug("Shutting down HttpPipeline")
-    parser.shutdownParser()
+    parser.synchronized {
+      isClosed = true
+      parser.shutdownParser()
+    }
     super.stageShutdown()
   }
 

